tensorstore/kvstore/s3/aws_metadata_credential_provider.cc

```diff
@@ -0,0 +1,179 @@
+// Copyright 2023 The TensorStore Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "aws_metadata_credential_provider.h"
+
+#include "tensorstore/internal/json_binding/absl_time.h"
+#include "tensorstore/internal/json_binding/bindable.h"
+#include "tensorstore/internal/json_binding/json_binding.h"
+
+using ::tensorstore::Result;
+using ::tensorstore::internal::ParseJson;
+using ::tensorstore::internal_http::HttpResponseCodeToStatus;
+using ::tensorstore::internal_kvstore_s3::AwsCredentials;
+
+namespace jb = tensorstore::internal_json_binding;
+
+namespace tensorstore {
+namespace internal_kvstore_s3 {
+
+namespace {
+
+// EC2 Metadata server url
+static constexpr char kEc2MetadataUrl[] = "http://http://169.254.169.254";
+// Token ttl header
+static constexpr char kTokenTtlHeader[] = "x-aws-ec2-metadata-token-ttl-seconds";
+// Token header
+static constexpr char kMetadataTokenHeader[] = "x-aws-ec2-metadata-token";
+
+static constexpr char kTokenPath[] = "/latest/api/token";
+static constexpr char kIamPath[] = "/latest/meta-data/iam/";
+static constexpr char kIamSecurityCredentialsPath[] = "/latest/meta-data/iam/security-credentials/";
+
+// Requests to the above server block outside AWS
+// Configure a timeout small enough not to degrade performance outside AWS
+// but large enough inside AWS
+static constexpr absl::Duration kConnectTimeout = absl::Milliseconds(200);
```

Comment:
include any directly named types.  I notice at least, in this file:

string
absl/time
absl/strings/str_cat
tensorstore/json_binding

Reply:
Done, I think the existing `tensorstore/internal/json_binding/*` imports covers the last case.

Reply:
There are a lot more includes missing.

 Look at the using declarations, for example. Anything that you spell out, such as 'tensorstore::Result' needs an import in each file where it is spelled (include-what-you-use); we don't rely on transitive includes for any of those cases. We only permit transitive includes for non-spelled field access and auto type use.

Reply:
Ah, thanks for pointing this out. I've gone through the file on a IWYU basis and adjusted the `#include`s accordingly.